sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+			receiver.indexedValue( object, "<collection element>", i );
+			i++;
+		}
+	}
+}
+----
+
+[NOTE]
+.Motivation for callback-style API
+====
+Instead of returning the extracted values from the method call, implementations of `ValueExtractor` pass the extracted values to the given receiver callback. This helps to avoid object allocations and allows to handle the case of a single extracted value (`Optional<T>`) and multiple extracted values (`List<T>`) in a uniform fashion.
+====
+
+==== Default extractors
+
+Compatible implementations provide extractors for the following types out of the box:
+
+* arrays of objects and all primitive data types
+* `java.util.Iterable` (when `@Valid` is given on the iterable element itself, the element and all its entries will be validated; this is to grant backwards compatability with BV 1.1; when `@Valid` is given on the type parameter of an iterable element, all the entries will be validated). A node of type `TYPE_PARAMETER` and with the name "<collection element>" will be added to the path when validating a type parameter constraint.
+* `java.util.Map` (when `@Valid` is given on the map element itself, the element and all its values will be validated; this is to grant backwards compatability with BV 1.1; when `@Valid` is given on the key type parameter, only the keys will be validated; when `@Valid` is given on the value type parameter, only the values will be validated. A node of type `TYPE_PARAMETER` and with the name "<map value>" will be added to the path when validating a type parameter constraint on the map value. A node of type `TYPE_PARAMETER` and with the name "<map key>" will be added to the path when validating a type parameter constraint on the map key.
+TODO: format of key representations
+* `java.util.Optional`. No node will be appended to the path when validating type-level constraints on `Optional`. Constraints given on an element of type `Optional` apply to the wrapped value by default (see below).
+* `javafx.beans.observable.ObservableValue`. No node will be appended to the path when validating type-level constraints on `ObservableValue`. Constraints given on an element of type `ObservableValue` apply to the wrapped value by default (see below).
+
+==== Plugging in custom extractors
+
+Additional value extractors can be registered with the Bean Validation provider in the following ways:
+
+* Invoke the new method `Configuration#addValueExtractor(ValueExtractor<?>)` (to apply it on the validation factory level)
+* Invoke the new method `ValidatorContext#addValueExtractor(ValueExtractor<?>)` (to apply it for a single `Validator`)
+* Specify the fully qualified class name of one or several extractors in `META-INF/validation.xml`: 
+
+[source,xml]
+----
+<value-extractors>
+    <value-extractor>com.example.MyExtractor</value-extractor>
+</value-extractors>
+----
+
+==== Retrieval of extractors
+
+When detecting a type-level constraint or cascade, the applicable extractor is determined as follows:
```

Comment:
> I'm not sure I'm reading the rules correctly but can I have one extractor for `Map<key>` and one for `Map<Value>`? It looks like you're saying no.

Yes, you can. An extractor is defined for one specific type parameter.

> The corollary is does an extractor extracting from a subtype container override the extractors of the supertype container?

Yes, it does override. Consider this example:

    List<@Valid Order> orders;

When looking up the extractor, the algorithm will do:

* Narrow the candidate extractors down to those that are declared for type parameter `T` of type `List`; e.g. this would dismiss the extractors for map key and map value but keep the ones for iterable and list.
* Narrow the candidate extractors down to the one(s) for the most specific sub-type; e.g. this would dismiss the one for `Iterable` and keep the one for `List` as the latter is a subtype of `Iterable`
* If there is exactly one extractor remaining, use it. If there are more than one or zero remaining, raise an exception.